### ISSUE TITLE
Get the value of the tuned hyperparameter

### ISSUE THREAD

**Issue by Coorch** [pomonam/Self-Tuning-Networks#2]
Thanks for your great work!

I found in your work, the tuned hyperparameter like dropout, is used with batchsize. 
![image](https://user-images.githubusercontent.com/38720272/109450115-1b2b6580-7a85-11eb-8e2a-e587eaf5d0c9.png)

If I want to tune the hyper like weight decay, how could I get the value of the htensor?

**Comment by pomonam** [comment]
Thank you for your interest! I have to admit that tuning weight decay might be a little bit tricky with the current codebase, as it requires different training and validation objectives. But feel free to modify the codebase to make it work. 

To answer your specific question, you can use the same function get_perturbed_hyper, but with same_perturb_mb=True. This will use the same hyperparameter per batch.

**Comment by Coorch** [comment]
Thanks for your reply!
And could you tell me more about the difference between training and validation? I feel a little bit incomprehensible.

**Comment by pomonam** [comment]
Sure! Let's say you are tuning weight decay on linear regression. The training objective would be MSE + regularization and the validation objective would be MSE on the validation data. I think that step optimizer takes in a single criterion, which is used for both training and validation objectives. You need to separate this out so that the objective when training will add regularization while the objective in training doesn't add regularization.

**Comment by Coorch** [comment]
OK，I understand!
Thanks for you patience! 
Hope you have a nice day！

**Comment by Coorch** [comment]
@pomonam Hello, thanks for your help, I have achieved great performance on tuning weight decay. 
After running a lot of tests, I get some questions:
1. When I set the num_workers to 4 in the dataloader
```python
train_loader = torch.utils.data.DataLoader(train_set, batch_size=info["train_batch_size"],
                                               shuffle=True, num_workers=0,
                                               worker_init_fn=np.random.seed(info["data_seed"]))
```
the performance will drop from 78(num_workers=0) to 76(num_workers=4, just change the num_workers) on the cifar10 experiment.

2. In the code, you have set the random seed:
```python
torch.manual_seed(args.model_seed)
np.random.seed(args.model_seed)
random.seed(args.model_seed)
```
So that every experiment has the same result. If I don't set the seed, the results will change every time, but sometimes, I can get better performance in 1. So is it necessary to set the seed?

Hope you can give me some suggestions, I would be very grateful

**Comment by pomonam** [comment]
Hello, that's good to hear! 

Regarding the first question, I am not sure what might be the problem. I suspect it might be due to the fact that the code is using (1) a custom dataset and (2) `next_batch` function to load the dataset. Note that I only tested & conducted the experiments with `num_workers = 0`.

On the second question, in those cases, you can run the experiment with several different seeds and average the results, so that you can make sure the algorithm is robust to random seeds. In the paper, we used 5 different random seeds and averaged the performance to report the final value.

**Comment by Coorch** [comment]
OK, thanks for your answer! I will attempt to use different random seeds.
Hope you have a good day!

**Comment by Coorch** [comment]
@pomonam I'm sorry to bother you again. 
This week, I got some problem about the batchnorm2d.
I find in the conv2d and linear, weights and bias are separated:
```python
output = F.conv2d(inputs, self.general_weight, self.general_bias, padding=self.padding,
                          stride=self.stride, groups=self.groups, dilation=self.dilation)
        

linear_hyper = self.hyper_bottleneck(h_net)

hyper_weight = linear_hyper[:, :self.out_channels].unsqueeze(2).unsqueeze(2)
hyper_bias = linear_hyper[:, self.out_channels:]
response_out = F.conv2d(inputs, self.response_weight, padding=self.padding, stride=self.stride,
                        groups=self.groups, dilation=self.dilation)
response_out *= hyper_weight
if self.response_bias is not None:
    response_out += (hyper_bias * self.response_bias).unsqueeze(2).unsqueeze(2)
    output += response_out
```
But in bn, they are together
```python
output = F.batch_norm(
            input,
            # If buffers are not to be tracked, ensure that they won't be updated
            self.running_mean if not self.training or self.track_running_stats else None,
            self.running_var if not self.training or self.track_running_stats else None,
            self.general_weight, self.general_bias, bn_training, exponential_average_factor, self.eps)
```
when I calculate the response output, the response_weight and response_bias are both required
```python
response_out = F.batch_norm(
            input,
            # If buffers are not to be tracked, ensure that they won't be updated
            self.running_mean if not self.training or self.track_running_stats else None,
            self.running_var if not self.training or self.track_running_stats else None,
            self.response_weight, self.response_bias, bn_training, exponential_average_factor, self.eps)
        response_out *= hyper_weight
```
So, is it necessary to add hyperbias * response_bias like conv2d and linear?
```python
if self.response_bias is not None:
    response_out += (hyper_bias * self.response_bias).unsqueeze(2).unsqueeze(2)
    output += response_out
```
